### PR TITLE
feat(issues-platform): Start writing occurrences to platform

### LINKED ISSUE #43975: Perf Migration: Start writing perf issues to the issue platform
We'll need to start writing performance issues to the issue platform dataset. When we do this, we'll be dual writing for some period of time. Since both perf issue detection and issue platform create/update the issue and counts, we'll need to make sure that only one of these systems is actively updating counts at any one time, otherwise we'll end up double counting.

It might make sense to split up our concept of writing here: Writing the events to the issue platform dataset and creating/updating issues. We can dual write events, since perf issues currently are writing to transactions, and in issue platform will be writing to a new dataset. And we should have a switch that flips us between creating issues via transaction ingestion vs issue platform. Might make sense to make the switch per org so that we can test it in sentry and make sure the issues still look/render correctly.

We should make sure that we send the issue to the issue platform *after* we've written the transaction to nodestore.

### PR DESCRIPTION
**Pull request by udameli**
Use system and project options set in https://github.com/getsentry/sentry/pull/45709

Fixes https://github.com/getsentry/sentry/issues/43975

### CODE REVIEW

File: src/sentry/event_manager.py
Comment (udameli) [inline src/sentry/event_manager.py:2486]: perf issues have `value` field in metadata

**Comment by barkbarkimashark** [comment]
I know this bit of code is where we wanna avoid double-counting issue stats:
https://github.com/getsentry/sentry/blob/149907ee5df4b4513db2e813357518d477859bcb/src/sentry/issues/ingest.py#L61-L63

Gonna take a quick look for any others.

**Comment by barkbarkimashark** [comment]
Hmm, not too sure if these TSDB models are used at all:
https://github.com/getsentry/sentry/blob/5bb33925a92b441d18a5c3aa84df2c37777cb395/src/sentry/event_manager.py#L2491

Need to double check whether we're actually querying this since we're not calling this function at all in issue platform.

**Comment by udameli** [comment]
@barkbarkimashark I've also updated `send_issue_occurrence_to_eventstream` to skip consume if it's a perf issue with options enabled and added another check [here](https://github.com/getsentry/sentry/blob/5fd8242aca575c84314f1b6744be2ec1d2e9e471/src/sentry/tasks/post_process.py#L529-L532)

File: src/sentry/tasks/post_process.py
Comment (wedamija) [inline src/sentry/tasks/post_process.py:552]: I think we want to remove this check. Passing `skip_consume=True` to `eventstream` means that we'll already skip sending these to `post_process_group`. But if we have this check here, then issue alerts will not fire for performance issues at all

File: src/sentry/issues/ingest.py
Comment (wedamija) [inline src/sentry/issues/ingest.py:57]: I think that ideally we'd skip calling this as well when the option isn't enabled. We need to do this to avoid calling https://github.com/getsentry/sentry/blob/2ec9adaa218e4097caa36e456c160b7cbe906b94/src/sentry/issues/ingest.py#L204 which will update counts on the issue and cause double counting.

I wonder if the best option here might be to add logic into `save_issue_from_occurrence` that just fetches the group and returns it when the option is disabled. If the group doesn't exist, then it just returns None so that we don't process it any further. 

I guess there should never be a case where the group doesn't exist since it'd have already been created by the transaction, but probably worth just being safe?

File: src/sentry/event_manager.py
Comment (wedamija) [inline src/sentry/event_manager.py:2505]: Do we want the span ids to be displayed by default, or are they just used by the UI to render things?

If we just use them as data, then it might make sense to include these in `evidence_data` instead.

Screen Shot 2023-03-14 at 9 19 51 AM
Screen Shot 2023-03-14 at 9 23 50 AM

Based on the UI and email^, I think we want to have the actual span details in the `evidence_display`, and then the span ids in `evidence_data`

Comment (wedamija) [reply src/sentry/event_manager.py:2505]: Although I'm not sure how generalizable that is to all perf issue types. But ideally, `evidence_display` has things we can display to the user, so that we can include them in email/slack/etc without having to customize for every new issue type.

Comment (udameli) [reply src/sentry/event_manager.py]: that makes sense. I update both evidence sections. I'm passing span ids because we look their descriptions on the client side but maybe that's not ideal

File: src/sentry/issues/ingest.py
Comment (udameli) [reply src/sentry/issues/ingest.py:57]: that makes sense. I've updated `save_issue_from_occurrence`

File: src/sentry/tasks/post_process.py
Comment (udameli) [reply src/sentry/tasks/post_process.py]: ah okay. removed!

File: src/sentry/event_manager.py
Comment (wedamija) [reply src/sentry/event_manager.py]: I'm wondering here if we should automatically translate the span ids to the span contents (just for `evidence_display`). That way it'll automatically display correctly. Then anything the FE client code needs to do with span ids can just work from `evidence_data`. I'm not totally sure if that makes sense?

This could be a thing where maybe we add `parent_spans, cause_spans, offender_spans` to the `PerformanceProblem`? And then we extract those from the transaction in the detector.
Since in the future, we won't actually be storing `PerformanceProblem` in nodestore at all, it'll just be used as an interface for you to send things to the issue platform.

Or maybe we should just emit `evidence_data` and `evidence_display` explicitly from the detectors. Then we'd be able to just shove it straight into evidence and it'd just work. I'm not sure if that makes sense here, but it means that each performance detector could more easily customize what they want to display.

File: src/sentry/issues/ingest.py
Comment (wedamija) [inline src/sentry/issues/ingest.py:66]: Feels like we've repeating this a fair bit, does it make sense to wrap it in a utility function somewhere?

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py]: yeah I think `evidence_data` and `evidence_display` should be explicitly emitted from the detectors. So that I don't have to translate PerformanceProblem output into evidence because it'd be different for each detector. I'm going to add that change in 👍

File: src/sentry/utils/performance_issues/performance_problem.py
Comment (wedamija) [inline src/sentry/utils/performance_issues/performance_problem.py:20]: Just to add to the TODO: We can't make it required until we stop loading these from nodestore via `EventPerformanceProblem`, since there's legacy data in there that won't have these fields. So until we disable transaction based perf issues we'll need to keep this optional.

File: src/sentry/utils/performance_issues/performance_problem.py
Comment (wedamija) [inline src/sentry/utils/performance_issues/performance_problem.py:56]: We should do this via `data.get("evidence_data", [])` so that it doesn't break for existing perf problems we're fetching from nodestore.

File: src/sentry/event_manager.py
Comment (wedamija) [inline src/sentry/event_manager.py:2488]: Looks like I might have missed this as part of our ingestion schema. What does `metadata["value"]` get used for?

Comment (udameli) [reply src/sentry/event_manager.py:2488]: it's used in the group stream. we generally show a query there 
image

File: src/sentry/utils/performance_issues/performance_problem.py
Comment (udameli) [reply src/sentry/utils/performance_issues/performance_problem.py:56]: ah 👍 that might fix these 10000 failing tests

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py:2488]: it's all the `SELECT`s and `GET /api/` in the screenshot

File: src/sentry/utils/performance_issues/performance_problem.py
Comment (udameli) [reply src/sentry/utils/performance_issues/performance_problem.py]: added!

File: src/sentry/utils/performance_issues/performance_problem.py
Comment (udameli) [reply src/sentry/utils/performance_issues/performance_problem.py]: updated!

File: src/sentry/event_manager.py
Comment (wedamija) [reply src/sentry/event_manager.py:2488]: Ok, looks like I missed this into the `IssueOccurrence` model. Right now we have `title` - `N+1 Api Call` and `subtitle` - `GET /api/...`. But we don't have anything for location, which is currently extracted from the event.

This all happens here: https://github.com/getsentry/sentry/blob/b134ebb0d4cbe3d5a03cd633d4bbb1c29cbd8545/src/sentry/issues/ingest.py#L121-L139

Feels to me like `subtitle` should map to `metadata["value"]`, and we should have a new `location` value that maps to `metadata["location"]`, and we fall back to `event["location"]` if not present. Does that make sense to you?

Comment (udameli) [reply src/sentry/event_manager.py:2488]: this is currently where the values are coming from in the UI 
Screenshot 2023-03-14 at 5 53 53 PM

for title, if there's no title in metadata, we'd default to pulling title from the group which is fine
for culprit, we read it from the group

so we don't really use location in this specific case

Comment (udameli) [reply src/sentry/event_manager.py:2488]: it seems that in the UI we don't really use `subtitle` that's sent https://github.com/getsentry/sentry/blob/master/static/app/utils/events.tsx#L112

Comment (udameli) [reply src/sentry/event_manager.py:2488]: > Feels to me like subtitle should map to metadata["value"], and we should have a new location value that maps to metadata["location"], and we fall back to event["location"] if not present. Does that make sense to you?

I think this can work. we're flipping culprit and metadata.value since the section right of the title is called a subtitle and the string below is called message but it should be fine I think

Comment (wedamija) [reply src/sentry/event_manager.py:2488]: Ahh interesting. It looks like `culprit` just happens to be the same as `metadata["location"]` for perf issues, so I made that assumption.

Ok, so I think that we should just make `culprit` explicit as part of the occurrence then. So
 - `title` - `metadata["title"]`
 - `subtitle` - `metadata["value"]`
 - `culprit` - `culprit`

Comment (udameli) [reply src/sentry/event_manager.py]: yeah there are some fields that I don't think we use ever. 
👍 that works! I updated subtitle and can follow up with more changes after this PR

Comment (wedamija) [reply src/sentry/event_manager.py]: https://github.com/getsentry/sentry/pull/45809

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py]: going back to this, we might need to go back to passing span ids in evidence display because we do some extra rendering customization like turning transaction name into a link and highlighting parameters in the url for http span description. adding some markup like that here is pretty cumbersome so I wonder if we need to go back to passing span ids and letting the client render the evidence sections as it needs
Screenshot 2023-03-15 at 1 56 18 PM
Screenshot 2023-03-15 at 1 56 52 PM

Comment (wedamija) [reply src/sentry/event_manager.py]: That's all basically UI customization, so in that case you're better off using the span ids from `evidence_data` and ignoring the `evidence_display` entirely. Then `evidence_display` can just be used for emails/slack (unless you also want to customize there)

File: src/sentry/event_manager.py
Comment (barkbarkimashark) [inline src/sentry/event_manager.py:2474]: Question about this. How long are we planning to dual-write here by project? I ask because the backfill script I'm working on has some baked in assumptions that we'll be fully dual-writing for all orgs and projects on a certain day.

Is the plan to fully cut-over performance issues to issue platform something like this:
1. Merge this PR to master.
2. Test out dual-writing for sentry org 
3. Remove this check so we can fully dual-write for all org-projects (this would be the date I need for the backfill script to avoid ingesting dupes)
4. Run backfill script
5. Cut-over search queries and stats to query on the issue platform dataset instead of transactions table

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py]: okay, then I'll keep everything as is 👍

File: src/sentry/issues/occurrence_consumer.py
Comment (udameli) [inline src/sentry/issues/occurrence_consumer.py:269]: forgot to add perf issues to this check

File: src/sentry/issues/ingest.py
Comment (udameli) [inline src/sentry/issues/ingest.py:80]: I've noticed that we were double hashing which is why occurrences were never saved. this fixes that

File: src/sentry/issues/occurrence_consumer.py
Comment (udameli) [reply src/sentry/issues/occurrence_consumer.py]: oh nevermind

**Comment by udameli** [comment]
I need to double check but somehow I see two issues created. Will report back

File: src/sentry/event_manager.py
Comment (wedamija) [inline src/sentry/event_manager.py:2505]: I wasn't sure whether we were using this producer correctly for a high volume environment, so asked in sns. Turns out that we aren't using it correctly, and flushing on every message will slow things down a fair bit.

Slack thread is [here](https://sentry.slack.com/archives/CLTE78L73/p1678931612520379), but I'll summarise a bit: 

 - It's slightly better to cache the `Producer` and not instantiate it every time
 - Flushing on every message is bad, since it makes what was an async call into a sync one and will slow down processing
 - We should consider using the Arroyo `Producer` since it returns futures for each flush that we can wait on.
 - We'll need to keep the futures from each `produce` call around somewhere so that we can be sure they complete. Recommendations here are to store the futures in a global, then use `atexit` to await the futures.

File: src/sentry/issues/producer.py
Comment (udameli) [reply src/sentry/issues/producer.py]: done

File: src/sentry/issues/producer.py
Comment (udameli) [reply src/sentry/issues/producer.py]: updated 👍

**Comment by barkbarkimashark** [comment]
> 

This issue is probably related to how we're doing search. Hopefully, this should fix it:
https://github.com/getsentry/sentry/pull/46016

File: src/sentry/issues/producer.py
Comment (barkbarkimashark) [inline src/sentry/issues/producer.py:38]: Do we need to gate this function for when `handle_occurrence_producer()` gets called on app exit? Is it possible futures keep getting appended to `occurrence_producer_futures` as the app is exiting?

Comment (wedamija) [reply src/sentry/issues/producer.py:38]: I don't think so, once `atexit` happens we're basically exiting the program, and so we shouldn't be producing anything extra.

File: src/sentry/issues/ingest.py
Comment (wedamija) [inline src/sentry/issues/ingest.py:80]: I feel like this should maybe be based on `skip_group_processing`. Otherwise, when we switch over to primarily create via the issue platform we won't hash the fingerprints here.

File: src/sentry/issues/utils.py
Comment (wedamija) [inline src/sentry/issues/utils.py:14]: Are these backwards? This function should return `True` when we should skip group processing, so should it be:
`is_performance and (not option_1 and not option_2)`?

I think that when this is `False` we also need to make sure that we're not creating groups via the old method.

Comment (udameli) [reply src/sentry/issues/utils.py:14]: yeah I think I need two separate flags here - to send occurrences to the platform (this will be tied to skipping group processing) and create a group. So we can first enable writing occurrences and then separately roll out group creation.

File: src/sentry/issues/producer.py
Comment (wedamija) [inline src/sentry/issues/producer.py:27]: Just a reminder for myself to look into these

File: src/sentry/issues/utils.py
Comment (wedamija) [inline src/sentry/issues/utils.py:10]: Why do we pass options in here instead of importing?

**Review by wedamija — COMMENTED** [review]
I think we're pretty close here. One thing we still need to do is skip these when `can_create_group` is True

https://github.com/getsentry/sentry/blob/a2dacb3cd7ef192b09c5ead84b6c8f12869a9fa4/src/sentry/event_manager.py#L2485
https://github.com/getsentry/sentry/blob/a2dacb3cd7ef192b09c5ead84b6c8f12869a9fa4/src/sentry/event_manager.py#L2487
https://github.com/getsentry/sentry/blob/a2dacb3cd7ef192b09c5ead84b6c8f12869a9fa4/src/sentry/event_manager.py#L2490

File: src/sentry/issues/utils.py
Comment (udameli) [reply src/sentry/issues/utils.py]: oh I thought passing in a dependency explicitly is better but yeah fixed it to use the import instead

File: src/sentry/event_manager.py
Comment (udameli) [inline src/sentry/event_manager.py:873]: not sure how this is going to work when there is a mix of n+1 db queries and other types in outputted performance problems. maybe it makes sense to turn group creation through platform on for all issue types at the same time

File: src/sentry/event_manager.py
Comment (udameli) [inline src/sentry/event_manager.py:1066]: same concern here

File: src/sentry/event_manager.py
Comment (wedamija) [reply src/sentry/event_manager.py:873]: It might be easier. Another option could be to build group creation into the policy layer, but I'm not sure if it's worth it since it's just temporary for this transition.

**Review by barkbarkimashark — CHANGES_REQUESTED** [review]
Please hold off on merging this until I rebuild the search_issues dataset with the correct primary key.

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py]: oh yeah you're right. fixed it!

File: src/sentry/event_manager.py
Comment (udameli) [reply src/sentry/event_manager.py]: > We don't need to filter like this in _get_or_create_release_many and _get_or_create_environment_many - transactions should create those regardless of whether they create groups. We just need to filter in _get_or_create_group_environment_many and _get_or_create_group_release_many.

oops, added correct checks

> It might actually be enough to just filter in _save_aggregate_performance, since those other functions operate only on the groups provided.

I've added a check just in case to be sure

Comment (udameli) [reply src/sentry/event_manager.py]: I was adding a check in the wrong place. resolving this since it's no longer applicable

File: src/sentry/issues/producer.py
Comment (udameli) [reply src/sentry/issues/producer.py:45]: added!

File: src/sentry/issues/ingest.py
Comment (wedamija) [inline src/sentry/issues/ingest.py:47]: `IssueOccurrenceData` is a dict, we'll need to handle that type in `can_create_group` as well

**Review by wedamija — COMMENTED** [review]
I think this is in a good state now, basically ready to merge once that last fix is in, and we've rebuilt the table.

File: src/sentry/issues/producer.py
Comment (wedamija) [reply src/sentry/issues/producer.py:27]: Checked with SNS, these config vals don't matter

File: src/sentry/event_manager.py
Comment (wedamija) [inline src/sentry/event_manager.py:1162]: I think this check is causing problems - We actually use `_get_or_create_group_release` in the issues ingest pipeline as well, and separately it's causing problems with errors. I think it's fine to rely on not adding the groups in `_save_aggregate_performance` instead, going to try removing

Comment (udameli) [reply src/sentry/event_manager.py]: oh that's what it was. yeah I saw all these failures

File: Makefile
Comment (mattgauntseo-sentry) [inline Makefile:135]: Not sure if you meant to keep this in, but if we we have this flag defined here: https://github.com/getsentry/sentry/blob/master/.github/actions/setup-sentry/action.yml#L77

Comment (wedamija) [reply Makefile:135]: We're debugging some stuff at the moment

**Review by wedamija — APPROVED** [review]
lgtm. @barkbarkimashark let us know when we're safe to merge here

**Review by barkbarkimashark — APPROVED** [review]
sorry, didn't look too closely at the actual code changes but we should be gtg for dual writes.

**Pull request merged**
